### PR TITLE
[Segment Replication] Fix flaky test testRelocateWhileContinuouslyIndexingAndWaitingForRefresh

### LINKED ISSUE #6531: [Segment Replication][BUG] Flaky Test SegmentReplicationRelocationIT.testRelocateWhileContinuouslyIndexingAndWaitingForRefresh
**Describe the bug**
When primary shard is relocated, as in this Integ test "testRelocateWhileContinuouslyIndexingAndWaitingForRefresh", on few occurrence old primary is trying to publish checkpoint after switching it's mode from primaryMode to replicaMode and fails. 

```
REPRODUCE WITH: ./gradlew ':server:internalClusterTest' --tests "org.opensearch.indices.replication.SegmentReplicationRelocationIT.testRelocateWhileContinuouslyIndexingAndWaitingForRefresh" -Dtests.seed=EF5EDA2370E498D8 -Dtests.security.manager=true -Dtests.jvm.argline="-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=64m" -Dtests.locale=sl -Dtests.timezone=Europe/Kyiv -Druntime.java=19

org.opensearch.indices.replication.SegmentReplicationRelocationIT > testRelocateWhileContinuouslyIndexingAndWaitingForRefresh FAILED
    com.carrotsearch.randomizedtesting.UncaughtExceptionError: Captured an uncaught exception in thread: Thread[id=117, name=opensearch[node_t0][refresh][T#1], state=RUNNABLE, group=TGRP-SegmentReplicationRelocationIT]
        at __randomizedtesting.SeedInfo.seed([EF5EDA2370E498D8:632A02144EF23820]:0)

        Caused by:
        java.lang.AssertionError: shard [test-idx-1][0], node[aRHMbFpHQ5icIAEeRaxYOQ], relocating [suhONszqTTaGlz_uwm_89w], [P], s[RELOCATING], a[id=hen83BYqQbSNbIsIwoVrfg, rId=XtIfsuxXTZmV3h-e4GoR1Q], expected_shard_size[208] is not a primary shard in primary mode
            at __randomizedtesting.SeedInfo.seed([EF5EDA2370E498D8]:0)
            at org.opensearch.index.shard.IndexShard.assertPrimaryMode(IndexShard.java:2325)
            at org.opensearch.index.shard.IndexShard.getReplicationGroup(IndexShard.java:3004)
            at org.opensearch.indices.replication.checkpoint.PublishCheckpointAction.publish(PublishCheckpointAction.java:121)
            at org.opensearch.indices.replication.checkpoint.SegmentReplicationCheckpointPublisher.publish(SegmentReplicationCheckpointPublisher.java:36)
            at org.opensearch.index.shard.CheckpointRefreshListener.afterRefresh(CheckpointRefreshListener.java:44)
            at org.apache.lucene.search.ReferenceManager.notifyRefreshListenersRefreshed(ReferenceManager.java:275)
            at org.apache.lucene.search.ReferenceManager.doMaybeRefresh(ReferenceManager.java:182)
            at org.apache.lucene.search.ReferenceManager.maybeRefreshBlocking(ReferenceManager.java:240)
            at org.opensearch.index.engine.InternalEngine$ExternalReaderManager.refreshIfNeeded(InternalEngine.java:432)
            at org.opensearch.index.engine.InternalEngine$ExternalReaderManager.refreshIfNeeded(InternalEngine.java:412)
            at org.apache.lucene.search.ReferenceManager.doMaybeRefresh(ReferenceManager.java:167)
            at org.apache.lucene.search.ReferenceManager.maybeRefresh(ReferenceManager.java:213)
            at org.opensearch.index.engine.InternalEngine.refresh(InternalEngine.java:1755)
            at org.opensearch.index.engine.InternalEngine.maybeRefresh(InternalEngine.java:1734)
            at org.opensearch.index.shard.IndexShard.scheduledRefresh(IndexShard.java:4064)
            at org.opensearch.index.IndexService.maybeRefreshEngine(IndexService.java:974)
            at org.opensearch.index.IndexService$AsyncRefreshTask.runInternal(IndexService.java:1107)
            at org.opensearch.common.util.concurrent.AbstractAsyncTask.run(AbstractAsyncTask.java:159)
            at org.opensearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:747)
            at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
            at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
            at java.base/java.lang.Thread.run(Thread.java:1589)
```

**To Reproduce**
Run SegmentReplicationRelocationIT.testRelocateWhileContinuouslyIndexingAndWaitingForRefresh for few hundred iterations to catch this flaky issue.
**Expected behavior**

**Plugins**
Please list all plugins currently enabled.

**Screenshots**

**Host/Environment (please complete the following information):**

**Additional context**

### PR DESCRIPTION
**Pull request by Rishikesh1159**
### Description
This PR fixes the flaky Test failing in `SegmentReplicationRelocationIT.testRelocateWhileContinuouslyIndexingAndWaitingForRefresh`. This PR is the best effort to reduce flakiness.

### Issues Resolved
#6531 

### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff
- [ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/indices/replication/checkpoint/PublishCheckpointAction.java [hunk 1]
```diff
@@ -118,7 +118,12 @@ final void publish(IndexShard indexShard) {
             PublishCheckpointRequest request = new PublishCheckpointRequest(indexShard.getLatestReplicationCheckpoint());
             final ReplicationCheckpoint checkpoint = request.getCheckpoint();
 
-            final List<ShardRouting> replicationTargets = indexShard.getReplicationGroup().getReplicationTargets();
+            final List<ShardRouting> replicationTargets;
+            try {
+                replicationTargets = indexShard.getReplicationGroup().getReplicationTargets();
```
Comment (dreamer-89) [inline server/src/main/java/org/opensearch/indices/replication/checkpoint/PublishCheckpointAction.java:123]: @Rishikesh1159 : Thanks for fixing this. Can you also please add more details around why this fix, reduces/fixes the flakyness here.

Comment (Rishikesh1159) [reply server/src/main/java/org/opensearch/indices/replication/checkpoint/PublishCheckpointAction.java:123]: Sure. During the process of relocation of primary, primary still publishes checkpoints on refresh even if relocation is in progress. So, during relocation when handoff is completed we switch the primarymode to false. So, this handoff process and publishing checkpoints to replicas happen parallely in our scenario.

Before publishing checkpoints to replicas we do call getReplicationGroup() on IndexShard. And this methods asserts on primarymode. So, very rarely we end up in a situation where primarymode of shard is switched between checkpointRefreshListeners afterRefresh check [here](https://github.com/opensearch-project/OpenSearch/blob/b0ab550ecfb80a2942b58169c5c38e500950a8cc/server/src/main/java/org/opensearch/index/shard/CheckpointRefreshListener.java#L43) and this getReplicationGroup() assert. 

By throwing an exception in getReplicationGroup() when shard not in primary mode we are reducing flakiness. But again we immediately have assert after this primarymode check in getReplicationGroup(), If relocation handoff completes between these two lines and primarymode switches, we may again fail assert, so there is flakiness but will not happen often.

We are not concerned about primary sending checkpoints to replica here, we are only trying to reduce flakiness.

**Comment by mch2** [comment]
@Rishikesh1159 With the revert of the wait_until change, and the decision to not support that behavior with segrep, I think we should revert https://github.com/opensearch-project/OpenSearch/pull/6366 making this change obsolete.  With that revert - https://github.com/opensearch-project/OpenSearch/pull/6637 is the only change we will need to release any wait_until reqs from nrt replicas.

**Comment by Rishikesh1159** [comment]
> @Rishikesh1159 With the revert of the wait_until change, and the decision to not support that behavior with segrep, I think we should revert #6366 making this change obsolete. With that revert - #6637 is the only change we will need to release any wait_until reqs from nrt replicas.

Sure @mch2, yes, if we revert changes made to PublishCheckpointAction class in #6366 then this PR and issue becomes obsolete. I will revert changes made to PublishCheckpointAction class by #6366, in PR : #6637

**Pull request closed without merging**